sensor/common/registry/registry_store.go

```diff
@@ -392,3 +392,233 @@ func (rs *Store) GetMatchingCentralRegistryIntegrations(imgName *storage.ImageNa
+			// If delegated scanning is disabled then we do not store additional secrets outside of those needed
+			// for scanning images from the OCP internal registry.
+			continue
+		}
+
+		if serviceAcctName != "" {
+			// Ignore secrets for service accounts other than default so that the
+			// default registry is not overwritten in the store.
+			continue
+		}
+
+		var err error
+		if isGlobalPullSecret {
+			err = rs.upsertGlobalRegistry(registryAddr, dce)
+		} else {
+			err = rs.upsertRegistry(namespace, registryAddr, dce)
+		}
+		if err != nil {
+			log.Errorf("unable to upsert registry %q into store: %v", registryAddr, err)
+		}
+	}
+}
+
+func (rs *Store) upsertSecretByName(namespace, secretName string, dockerConfig config.DockerConfig, serviceAcctName string) {
+	isGlobalPullSecret := namespace == openshiftConfigNamespace && secretName == openshiftConfigPullSecret
+
+	// hasBoundServiceAccount indicates that this secret is bound to a service account,
+	// which means the secret is managed by OCP and its lifecycle is tied to that of
+	// the associated service account (ie: if the service account is deleted so is the secret).
+	hasBoundServiceAccount := serviceAcctName != ""
+
+	// To avoid partial upserts - hold the lock until the entire secret upserted.
+	rs.storeMutux.Lock()
+	defer rs.storeMutux.Unlock()
+
+	for registryAddress, dce := range dockerConfig {
+		registryAddr := strings.TrimSpace(registryAddress)
+
+		if hasBoundServiceAccount {
+			// Registries found in any `dockercfg` secret bound a service account
+			// are assumed to reference the OCP internal registry.
+			rs.upsertPullSecretByNameNoLock(namespace, secretName, registryAddr, dce)
+			rs.addClusterLocalRegistryHost(registryAddr)
+			continue
+		}
+
+		if env.DelegatedScanningDisabled.BooleanSetting() {
+			// If delegated scanning is disabled then we do not store additional secrets outside of those needed
+			// for scanning images from the OCP internal registry.
+			continue
+		}
+
+		if isGlobalPullSecret {
+			if err := rs.upsertGlobalRegistry(registryAddr, dce); err != nil {
+				log.Errorf("Upserting global registry for pull secret %q, namespace %q, address %q: %v", secretName, namespace, registryAddr, err)
+			}
+		}
+
+		// Regardless if this secret is the global pull secret, we still store it
+		// in case there is a workload that directly references it by name.
+		rs.upsertPullSecretByNameNoLock(namespace, secretName, registryAddr, dce)
+	}
+
+	log.Debugf("Upserted %d entries from secret %q in namespace %q", len(dockerConfig), secretName, namespace)
+}
+
+func (rs *Store) upsertPullSecretByNameNoLock(namespace, secretName, registryAddr string, dce config.DockerConfigEntry) {
+	registryAddr = urlfmt.TrimHTTPPrefixes(registryAddr)
+
+	name := genIntegrationName(pullSecretNamePrefix, namespace, secretName, registryAddr)
+	ii := createImageIntegration(registryAddr, dce, name)
+
+	reg, err := rs.factory.CreateRegistry(ii, types.WithGCPTokenManager(gcp.Singleton()))
```

Comment:
seems like this is created unconditionally while the older method does it on-demand. Curious why this approach was chosen

Reply:
Not sure I follow, on upsert for both the new or old way the registry is created on demand.  The old way that creation was abstracted behind the [Set's UpdateImageIntegration](https://github.com/stackrox/stackrox/blob/f6627c14c7fd6b38213b506d688ef3efaea88a34/pkg/registries/set_impl.go#L160) - or am I missing something?

Reply:
I don't even remember unfortunately, so I guess we can close this